### ISSUE TITLE
Ankidroid crashes when installed alongside Tor Browser

### ISSUE THREAD

**Issue by nvbln** [ankidroid/Anki-Android#5337]
###### Reproduction Steps

1. Install Tor Browser alongside Ankidroid.
2. Either use Tor Browser or not, it doesn't seem to matter.
3. Open Ankidroid

###### Expected Result
It works.

###### Actual Result
It crashes.

###### Debug info
```06-20 16:44:41.031  1262  5962 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10008000 cmp=com.ichi2.anki/.DeckPicker} from uid 10092
06-20 16:44:41.031  1262  5962 W ActivityManager: Activity is launching as a new task, so cancelling activity result.
06-20 16:44:41.038  1262  1274 W ActivityManager: Duplicate finish request for ActivityRecord{89922f1 u0 com.ichi2.anki/.IntentHandler t1114 f}
06-20 16:44:41.047 28104 28104 W ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@7438cda
06-20 16:44:41.052 28104 28104 I AnkiDroid: openCollection: /storage/emulated/0/AnkiDroid/collection.anki2
06-20 16:44:41.094 25960 25960 W kworker/0:3: type=1400 audit(0.0:2352): avc: denied { search } for name="/" dev="tmpfs" ino=18539 scontext=u:r:kernel:s0 tcontext=u:object_r:tmpfs:s0 tclass=dir permissive
=0
06-20 16:44:41.132 28104 28120 D NetworkSecurityConfig: No Network Security Config specified, using platform default
06-20 16:44:41.183 28104 28104 D OpenGLRenderer: Skia GL Pipeline
06-20 16:44:41.199  1262  5962 W ActivityManager: Background start not allowed: service Intent { act=start-gecko cmp=org.torproject.torbrowser/org.mozilla.gecko.GeckoService (has extras) } to org.torproje
ct.torbrowser/org.mozilla.gecko.GeckoService from pid=28052 uid=10105 pkg=org.torproject.torbrowser startFg?=false
06-20 16:44:41.200 28104 28104 D AndroidRuntime: Shutting down VM
06-20 16:44:41.200 28104 28104 E AndroidRuntime: FATAL EXCEPTION: main
06-20 16:44:41.200 28104 28104 E AndroidRuntime: Process: com.ichi2.anki, PID: 28104
06-20 16:44:41.200 28104 28104 E AndroidRuntime: java.lang.IllegalStateException: Not allowed to start service Intent { act=start-gecko cmp=org.torproject.torbrowser/org.mozilla.gecko.GeckoService (has ex
tras) }: app is in background uid UidRecord{116949b u0a105 CACC idle change:cached procs:1 seq(0,0,0)}
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.os.Parcel.createException(Parcel.java:1958)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.os.Parcel.readException(Parcel.java:1918)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.os.Parcel.readException(Parcel.java:1868)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.support.customtabs.ICustomTabsService$Stub$Proxy.warmup(ICustomTabsService.java:162)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.support.customtabs.CustomTabsClient.warmup(CustomTabsClient.java:168)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at com.ichi2.compat.customtabs.CustomTabActivityHelper.onServiceConnected(CustomTabActivityHelper.java:123)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at com.ichi2.compat.customtabs.ServiceConnection.onCustomTabsServiceConnected(ServiceConnection.java:38)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.support.customtabs.CustomTabsServiceConnection.onServiceConnected(CustomTabsServiceConnection.java:32)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.app.LoadedApk$ServiceDispatcher.doConnected(LoadedApk.java:1730)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.app.LoadedApk$ServiceDispatcher$RunConnection.run(LoadedApk.java:1762)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.os.Handler.handleCallback(Handler.java:873)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:99)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:193)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:6718)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
06-20 16:44:41.200 28104 28104 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

###### Research
*Enter an [ x ] character to confirm the points below:*

[ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[  ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x ] I have searched for similar existing issues here and on the user forum

**Comment by mikehardy** [comment]
I like the idea of Tor so I spent time really investigating this

This is an upstream issue. TorBrowser is "[doing it wrong](https://developer.android.com/about/versions/oreo/android-8.0-changes.html#back-all)"

The [mozilla project they are based on has fixed it already](https://trac.torproject.org/projects/tor/ticket/29238#comment:3).

The tor project itself already worked on this type issue: https://trac.torproject.org/projects/tor/ticket/29238

I logged this corresponding upstream issue for them: https://trac.torproject.org/projects/tor/ticket/30944#ticket

Workaround is to not use Tor Browser as your default webview until they conform to Android 8 background limits? (I'm assuming that's your setup?)

Others have seen this issue and apparently client apps can [ignore the exception](https://github.com/openid/AppAuth-Android/issues/425#issuecomment-445315136) that happens when apps do it wrong, which is an easy change for us - #5338 

Please re-test on the next alpha after #5338 is merged, assuming it passes examination by @timrae

**Comment by nvbln** [comment]
Thanks a lot for the quick reply (and the extensive research)!

> Workaround is to not use Tor Browser as your default webview until they conform to Android 8 background limits? (I'm assuming that's your setup?)

I did not have it set as the default webview, rather I have no default webview to begin with. However, once I do set a default webview that is not the Tor Browser the problem is indeed fixed.

**Comment by mikehardy** [comment]
Cool - always nice when a diagnosis seems on track to be correct. I still like the idea of Tor so I'm curious about the results of #5338 - I'll leave this open till then
